### ISSUE TITLE
Feature Request: layer2: Add config parameters to allow layer2-speaker to run out-of-cluster

### ISSUE THREAD

**Issue by chrigl** [metallb/metallb#732]
Hey metallb-folks,

I am well aware of all the other l2 changes in flight at the moment, and certainly will rework this PR once the other stuff lands in main. But first, I want to know if this is of general interest for the project.

I would like to be able to manually configure mlPeers to be able to run metallb-speaker (l2) out-of-cluster. To illustrate what would be needed at the current point in time, I created a PR in my fork, so that you can get an idea of the changes: https://github.com/chrigl/metallb/pull/1/files

The rational behind this. Imagine a black box (or rack) with two switches (or cumulus routers) and a couple of kubernetes nodes. The switches are not part of the kubernetes cluster. Each node connects to both switches and bgp announce a node IP, as well as service, LB and pods ips to the switches. This setup tolerates losing one switch without affecting much functionality.

The switches have (in some setups) a l2-only connection to a customer network each, with no way to bgp peer to the customer network. We now want to provide services to the customer in an HA way, this is where metallb-speaker in l2 mode comes into play. Until now, we used metallb-speaker 0.6.2, with kubernetes leader election, so that always one switch announces the LB addresses. This worked well until kubernetes 1.18, which breaks client-go in the old version of metallb 0.6.2.

Long story short. By adding ml-peers to manually provide static ip addresses for MemberList (rather than using PodIPs), it is possible again to run metallb-speaker standalone on the switches. Even better, because how the announcement works, load might be splitted across both switches (what was not possible with bare leader election).

Please let me know what you think about this idea.
Big thanks for all your work :)

Example

```
                   ~~~~~~~~ 
                 /          \
                |  customer  | (layer 2)
                 \          /
                   ~~~~~~~~
                  |       |
         +----------+    +----------+
         | router 1 |    | router 2 | (do not belong to the kubernetes cluster)
         +----------+    +----------+
         | speaker~~~~~~~~~speaker  | (cfg MemberList + ignore endpoint-on-node-patch
         |          |    |          |  speaker are connected to kube-apiserver)
         |          |    |          | 
+~~~~~~~~~~bgp      |    | bgp~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
|        +----------+    +----------+                     |
|          |         \  /         |                       | bgp announcement:
|          |          \/          | fixed, but            |  Node IPs (the one kubelet uses for node-ip)
|          |          /\          | internal, ips.        |  Pod IPs
|          |e0     e1/  \e0     e1|                       |  Service IPs
|        +----------+    +------------------------+       |  LoadBalancer IPs
|        | worker 1 |    | worker 2               |       |
|        +----------+    +------------------------+       |
+~~~~~~~~~~~bgp     |    |  bgp~~~~~~~~~~~~~~~~~~~~~~~~~~~+
         |  kubelet |    |  kubelet               |
         |  node_ip |    |  node_ip (on loopback) |
         +----------+    +------------------------+
         | kube-api |    | svc/ingress            | (e.g. 192.168.178.1)
         +----------+    +------------------------+
```

* each kubernetes worker is connected to both routers
* bgp daemon on kubernetes worker are connected to both routers (configured with ECMP)
* each router knows how to reach LoadBalancer IPs
* one metallb is responsible for 192.168.178.1
* if the responsible metallb-speaker detects a who-has request, it answers with its own mac
* if the responsible metallb-speaker fails, the remaining takes over

**Comment by johananl** [comment]
Hey @chrigl. I have to admit I'm struggling to understand the use case. I'm wondering if a diagram would help.

Anyway, did I get it right that the main idea is being able to run a MetalLB speaker process outside a k8s cluster? If so, right now MetalLB has exactly one deployment target, and that is a k8s cluster. In other words, we aren't testing or supporting running MetalLB on platforms which aren't k8s. On the surface, I'd say that if there is demand in the community for such an option we can definitely discuss it, however from the info I have right now your use case sounds highly specialized, which makes it tricky for us to solve in the generic MetalLB codebase.

**Comment by russellb** [comment]
Even if the memberlist peer IPs were configurable, I think the logic in the l2 speaker that chooses which speaker owns each IP is broken.  It only chooses the speaker that's running on a node along with one of the service endpoints.  That would be totally busted in this case, right?  I guess that's similar to #646 or #717 

I tend to agree that this really isn't a supported use case.  Even if we added a knob to make it work again right now, it's highly likely to get broken again.  I tend toward suggesting maintaining this as a custom fork.

**Comment by chrigl** [comment]
Thanks for your feedback. I highly appreciate it!

Added a diagram to the ticket description. Maybe this clarifies the setup.

I totally relate to both of your answers. Yes, it is highly specialised, and yes, because it is a not supported use case, it might break again.

> Even if the memberlist peer IPs were configurable, I think the logic in the l2 speaker that chooses which speaker owns each IP is broken. It only chooses the speaker that's running on a node along with one of the service endpoints. That would be totally busted in this case, right? I guess that's similar to #646 or #717

The patch would include, that in out-of-cluster use case, the Kubernetes Nodes will not be part of the evaluation in `usableNodes`. Selecting the actual route in this case is part of another layer.

I also totally relate to the suggestion of maintaining it as a custom fork. I started with the idea of just copying over some of the metallb l2 files, until I noticed, that I can reuse almost everything and just need a couple of lines changed. This was when I thought, it might be worth to ask here :)

**Comment by johananl** [comment]
@chrigl I took a look at https://github.com/chrigl/metallb/pull/1/files. Unless we see that there is a significant demand for this functionality (i.e. it's a common use case), I'm hesitant to add this functionality despite the fact that it isn't a lot of code. Specifying a static list of node addresses doesn't suit the current design of MetalLB, and in addition I am worried about affecting the common layer 2 use case negatively and/or making it harder to maintain the layer 2 code.

I'm sure you understand that we have to somehow keep MetalLB focused on generic use cases. I'm fine with leaving this issue open to see if more users would like to see this implemented, however right now it doesn't look to me like something we want to merge.
